### DISCUSSION TITLE
my images at sun up or down have a dark band to one side, is there a way to solve that with a bat script?

### DISCUSSION THREAD

**Discussion by TLmv** [ImageMagick/ImageMagick#6432] [Help]
### ImageMagick version

7.1.1

### Operating system, version and so on

WIN 10 pro

### Description

I make a lot of pictures for time lapse, but due to lens angle at sun up or down i have a visible darker band on the opposite side at those moments. 
This band will be smaller and less distinct as the sun goes higher.

It has to do with the curve in the lens and light cominf from the side that from a certain point more light reflects off, making that part darker, at widest it's about 1/5 of the frame width.

By using setting the mean from snibgo i tried gain and neggain with use of reference frames to try and smoothen that dark vertical band out, this did have some effect. 
but not enough for my liking, the dark side is still visible darker.

by now I have this bat script that tells me via the reference frames if there is a dark side and if it's left or right. but i got no further in trying to smooth this out.

is this possible at all or am i trying to solve something that can't be fixed automatically?

**Comment by snibgo** [comment]
Please show a sample image.

**Comment by TLmv** [comment]
on a singular, resized image it's a bit harder to spot.
so i made a stitched version of 2 images, this way it stands out more.
in this case the dark band is the right side of the left image....and less visible also on the right side of the right image.

**Comment by snibgo** [comment]
Were the two photos made at exactly the same exposure? Or was the camera on auto-exposure? Did you stitch from unedited raw files, or from out-of-camera jpegs? Did you use physical filters, eg polarizers?

I assume the effect is visible only in the sky. Is that correct?

The sky isn't a constant colour, of course. It gets darker as we turn away from the sun.

Anyhow, yes, if the amount of darkening is the same at all heights in the sky, we can correct quite easily. Suppose the pixel at top-left is sRGB(20%,30%,40%), and the pixel at the top of some other column is sRGB(17%,25%,33%). Then we can multiply that column by (20/17,30/25,40/33) to make that pixel exactly match the top-left pixel.

We can do that arithmetic by scaling the top-left pixel to the full width and height. Call that ImgA. Then crop the top row and scale that to the full height, and call that ImgB. Multiply the input image by ImgA/ImgB.

Your image is noisy (I *HATE* jpgs as input files), so instead of taking the top row, I take the average of the top 100 rows.

Windows BAT script, where `bandedSky.jpg` is your stitched image. This needs IM to be HDRI.
```
magick ^
  bandedSky.jpg +write mpr:INP ^
  -define compose:clamp=off ^
  -set option:MYSIZE %%[fx:w]x%%[fx:h] ^
  ( mpr:INP ^
    -crop 1x1+0+0 +repage ^
    -scale "%%[MYSIZE]^!" ^
  ) ^
  -compose Multiply -composite ^
  ( mpr:INP ^
    -crop x100+0+0 +repage ^
    -scale "x1^!" ^
    -scale "%%[MYSIZE]^!" ^
  ) ^
  -compose DivideSrc -composite ^
  out.png
```

[Image removed at OP's request.]

We might need to make a mask of the sky, so the effect is applied just there.

**Comment by TLmv** [comment]
the result you posted looks good, maybe a bit bright at the far edge of the image, but can't spot the middle dark band and that was my main focus.

it's an interesting approach that now i see it is something i half thought up myself and tried, making everything much too bright before abandoning that idea,

and yes it's mostly visible in the lighter parts of the image e.g. the sky. exposure as you might have guessed is not always identical, light might not be polarized, but is filtered and sadly out of camera jpg :(

i don't like jpg either, but for this project i had to accept the noise and some other things due to hardware and real world limitations.
but if i can work this out i'm sure i will use it on some better quality projects as well that i would be willing to show. 
so please, before i feel ashamed of posting a noisy jpg, would you remove this result image you posted.

will take me some days however to rewrite my script to work with this code you've given me, as soon as i've got it working i'll let you know how it turns out crunching through various lighting settings.

**Comment by snibgo** [comment]
Many variations are possible. For example: we adjust the left-most column to make the top-left pixel match the mean of the top 100 pixels in that column. We also adjust the right-most column to make the top-right pixel match the mean of the top 100 pixels in that column. We adjust all other columns linearly between those adjustments.

This means we keep lightness in the left, and darkness in the right.
```
magick ^
  bandedSky.jpg ^
  -alpha off ^
  +write mpr:INP ^
  -define compose:clamp=off ^
  -set option:MYSIZE %%[fx:w]x%%[fx:h] ^
  (  ^
    ( mpr:INP ^
      -crop 1x100+0+0 +repage ^
    ) ^
    ( mpr:INP ^
      -crop 1x100+%%[fx:w-1]+0 +repage ^
    ) ^
    +append +repage ^
    -scale "2x1^!" ^
    -resize "%%[MYSIZE]^!" ^
  ) ^
  -compose Multiply -composite ^
  ( mpr:INP ^
    -crop x100+0+0 +repage ^
    -scale "x1^!" ^
    -scale "%%[MYSIZE]^!" ^
  ) ^
  -compose DivideSrc -composite ^
  +depth ^
  x.png
```

**Comment by TLmv** [comment]
i am a bit struggling with command line, your code looks fine.
as far as i can tell, i have the HDRI version: ImageMagick-7.1.1-Q16-HDRI

after no succes, i restarted by copying your code and only added 

`cls`
`setlocal EnableDelayedExpansion`


and a filepath change; D:\input\bandedSky.jpg

it seems to close the terminal after -compose Multiply -composite
and alas i'm clueless what i'm missing....

**Comment by snibgo** [comment]
Please show your script. To ensure it formats correctly, add a line with three back-ticks, like \`\`\`, at the top of the script, and the same at the bottom.

EDIT: You say:
 > it seems to close the terminal after ...

That isn't helpful. IM won't close any terminal. Perhaps you are not running the script in a command window (aka console, aka terminal). It would help if you insert a `pause` command after the `magick` command.

**Comment by TLmv** [comment]
IM won't but errors do close the command window.
However, if i put everything on one line withou all those ^, it works.
i think it goes wrong due to the ^ between "", i know it shouldn't but that is my best guess.

 -scale "x1^!" ^

```
REM @echo off
cls
setlocal enabledelayedexpansion


magick ^
  D:\input\bandedSky.JPG +write mpr:INP ^
  -define compose:clamp=off ^
  -set option:MYSIZE %%[fx:w]x%%[fx:h] ^
  ( mpr:INP ^
    -crop 1x1+0+0 +repage ^
    -scale "%%[MYSIZE]^!" ^
  ) ^
  -compose Multiply -composite ^ 
  ( mpr:INP ^
    -crop x400+0+0 +repage ^
    -scale "x1^!" ^
    -scale "%%[MYSIZE]^!" ^
  ) ^
  -compose DivideSrc -composite ^
  J:\Y2023D0522\IMG_0001.png

ECHO done
PAUSE
endlocal

```

**Comment by snibgo** [comment]
Your line ...
```
  -compose Multiply -composite ^ 
```
... ends with a space *after* the caret `^`. Caret means "the command continues on the next line" only when it is the final character on the line. When it isn't the final character, it means "use the next character literally". Remove the space.

**Reply by TLmv** [reply]
i don't know how that could have happened.
but now you showed me i'm changing my editor to color the spaces...hopefully next time i won't be searching for the unseen ;)
in any case thank you very much, it reads a lot better when it's not a oneliner...and after some last tweaks, the script is now running and all i have to do is wait.

first impression is that red becomes a bit more visible with darker images and low light blue skys loose a lot of white from clouds in the images.
but the dark band is gone, that much is true. i think i need to tinker some with the reference crop.

i also set up the second one you posted to run after this...see what looks best. 
was expecting maybe that my GPU would be involved for this, but it seems to be CPU all the way

**Comment by snibgo** [comment]
I sometimes accidentally add a space after the final caret. That's how I knew what to check.

I always break long commands into logical lines. It makes development, debugging and maintenance much easier.

On GPU, see https://www.imagemagick.org/script/opencl.php. It is disabled by default. To enable it, run the Windows command ...
```
set MAGICK_OCL_DEVICE=true
```
... anytime before the `magick` command.

**Comment by TLmv** [comment]
it gives fairly good result on most images i tested this far, 
although, sometimes i see result that looks like it's raining due to switches in vertical line color and i have a few images in a series that look like it captured northern lights, with strong vertical glowing lines :)
and this may be very artistic and all, but I'm not sure i'm THAT much of an artist.
visually, although setmeangain did not solve the dark band, the results were more usable with less variation between images.

tried different height settings in  `-compose Multiply -composite` but that did not solve those vertical lines.

**Comment by snibgo** [comment]
Is that using my first (short) command, or the second (longer)?

> ... it's raining due to switches ...

A blur somewhere might fix that.

If you can post a sample input, or upload it to somewhere I can download it, I'll take a look.

**Comment by TLmv** [comment]
before you came with this idea i was actually only doing one image of the 2 before merger, while this idea involves retouching the total image after merger. however i do not believe running it on one half will create a better outcome.

i do not think a blur will solve it, but i'll show you the result of the first (short) command.
this might not be the images with the highest dark part to solve, but it has a few different results in 25 images, making it a good example where 906 is nicely done, but after it becomes less appealing.

if my script is working properly, it should be that this series takes the right side as starting point, so instead of crop 1x1+0+0, it's 1x1+8501+0, same goes for the crop line

**Comment by snibgo** [comment]
I suppose the files in directory `original` are the inputs, and files in directory `im_script` are the outputs.

I see what you mean about the "rain" in the outputs. But I can't reproduce the problem. Putting `original/Y2023D0414-IMG_0907.JPG` through my short script makes a smooth result.

**Comment by TLmv** [comment]
on this day i have come across much worse rain, bright and off color outputs with darker images. but i thought the ones i send you were more diverse in showing problems within a limited range of frames. 
maybe you tested with 1x1+0+0 instead of 1x1+8500+0. however, if you can't reproduce my im_script outputs with your short script, there is nothing to solve....probably some mistake in my code.

i have done some more tests and while i solved a few of the overbright ones, the color changes stay, 
and it still has the 'rain' plus the images come out a bit more blurry. furthermore, i found that on sunset or sunrise the band is actually more a V shape, being enhanced instead of reduced with this code.

thank you for the code and taking the time to check my results, i learned a lot. but as you can imagine, i won't pursue this idea further with these results.

**Comment by snibgo** [comment]
I tested with `1x1+0+0` and `1x1+8500+0` and the more general `1x1+%%[fx:w-1]+0`. They were all good. I suspect a bug in your script. If you paste it here, I'll take a look.

The inputs are lighter on the right than on the left. If you adjust to match the right side, the images will become brighter. Some values will become greater than 100%, and this will be clipped to 100% when saving to JPG or PNG. Some RGB channels will be clipped, and not others, and this will change hues.

So for these images I suggest using the left side, which will generally darken images, with no danger of clipping. My second (longer) script might also clip, but this is not likely.

You have multiple images of the same scene. If you are making a movie, you might normalise so all the images have the same colours in the top 100 (or whatever) pixels.

**Comment by TLmv** [comment]
at this moment i'm using 1 and 8500, and although slightly better and less problems with that range you got, at least 930 with me still look very special. but i will run it the other way 'round so it becomes darker.

the same color would be a bit much i think, since you wouldn't see it become light, but yes, like the originals successive shots are in the same color range so it would help if they could stay that way.

here `-compose Multiply -composite`  is set at 400px height, but also tried 100, 200 and 800 in other runs to see if i could get less color change.

```
set imagnr=930
set subfolder=D:\
REM set startpoint=1
set startpoint=8500
set /a padded_imagenr=10000+%imagenr%
set outputname=Y2023D0414

magick ^
  %subfolder%original\%outputname%-IMG_%padded_imagenr:~-4%.JPG +write mpr:INP ^
  -define compose:clamp=off ^
  -set option:MYSIZE %%[fx:w]x%%[fx:h] ^
  ( mpr:INP ^
    -crop 1x1+%startpoint%+0 +repage ^
    -scale "%%[MYSIZE]^!" ^
  ) ^
  -compose Multiply -composite ^
  ( mpr:INP ^
    -crop x400+%startpoint%+0 +repage ^
    -scale "x1^!" ^
    -scale "%%[MYSIZE]^!" ^
  ) ^
  -compose DivideSrc -composite ^
  %subfolder%IM_script\%outputname%_IMG_%padded_imagenr:~-4%.JPG
```

**Comment by snibgo** [comment]
There is a bug. You have:
```
magick ^
  %subfolder%original\%outputname%-IMG_%padded_imagenr:~-4%.JPG +write mpr:INP ^
  -define compose:clamp=off ^
  -set option:MYSIZE %%[fx:w]x%%[fx:h] ^
  ( mpr:INP ^
    -crop 1x1+%startpoint%+0 +repage ^      <== Correct
    -scale "%%[MYSIZE]^!" ^
  ) ^
  -compose Multiply -composite ^
  ( mpr:INP ^
    -crop x400+%startpoint%+0 +repage ^     <== BUG
    -scale "x1^!" ^
    -scale "%%[MYSIZE]^!" ^
  ) ^
  -compose DivideSrc -composite ^
  %subfolder%IM_script\%outputname%_IMG_%padded_imagenr:~-4%.JPG
```

The line I have marked as BUG should be:
```
    -crop x400+0+0 +repage ^
```
The offset should be zero whether you normalise to the top-left or top-right or anywhere else. When the offset isn't zero, the column is divided by the mean of some other column, which is wrong.

**Comment by TLmv** [comment]
ok, i changed that and also made the darkest side leading, it's running again.
but the dark side makes the sun unable to come up, it stayed a dark blue sky  until it is high enough for the dark band to almost dissapear and then it explodes into the day.

i made a small image to show.
on the background you see 2 consecutive frames: cloud at nearly the same moment but the color is definately off in one of them.
the front image is the 'rain', for some reason this shot was also much lighter than other shots around it.

**Comment by snibgo** [comment]
What are the two input images?

**Comment by TLmv** [comment]
by now i've analyzed the images, it seems that a few pixels on the edges are off. now i'm rerunning the script with the 1x1+0+0 as 1x1+5+0.
as it has just started it'll take some time, but i haven't spotted rain yet and haven't seen any extreme light differences until now.

still has some color changes though, but they seem less.
i'll have to wait until it's finished, but this looks promising.....perhaps i need to do 1x1+5+5 to solve a bit more of the color differences.

**Comment by snibgo** [comment]
For the sequence, the clouds cause problems. We can deal with that.

However, I have just got home from a cataract operation. I currently can't see properly or think properly. Perhaps I'll be okay tomorrow.

**Comment by TLmv** [comment]
i hope all is well.  and wear sunglasses for a few days, it helps ;)

as for the script, although it has improved, the results with clouds are indeed not stable.

and now the rain have become sunrays.
